src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>point list\<close> |
+  is_intersect_found: IntersectFound \<open>point\<close> \<open>block\<close> |
+  is_intersect_not_found: IntersectNotFound \<open>block\<close>
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Consumer" |
+  "agent_in_state' _ = Producer"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (
+    partial_case m of
+      FindIntersect _ \<Rightarrow> Intersect |
+      RequestNext \<Rightarrow> CanAwait)" |
+  "next_state' Intersect m = (
+    partial_case m of
+      IntersectFound _ _ \<Rightarrow> Idle |
+      IntersectNotFound _ \<Rightarrow> Idle)" |
+  "next_state' CanAwait m = (
+    partial_case m of
+      RollForward _ _ \<Rightarrow> Idle |
+      RollBackward _ _ \<Rightarrow> Idle |
+      AwaitReply \<Rightarrow> Idle)" \<comment> \<open>only for this initial implementation\<close>
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+definition rollback_to :: "chain \<Rightarrow> point \<Rightarrow> chain" where
+  [simp]: "rollback_to \<C> point = takeWhile (\<lambda>block. hash block \<le> point) \<C>"
+
+corec consumer_program where
+  "consumer_program params = (
+    let (\<C>, phase, candidate_points) = params in
+    case phase of
+      IntersectionFinding \<Rightarrow>
+        \<up> Cont (FindIntersect (candidate_points \<C>));
+        \<down> M; (partial_case M of
+          Cont (IntersectNotFound _) \<Rightarrow>
+            \<up> Done;
+            \<bottom> |
+          Cont (IntersectFound _ _) \<Rightarrow>
+            consumer_program (\<C>, ChainUpdate, candidate_points)
+        ) |
+      ChainUpdate \<Rightarrow>
+        \<up> Cont RequestNext;
+        \<down> M; (partial_case M of
+          Cont (RollForward header _) \<Rightarrow>
+            consumer_program ((\<C> @ [header]), phase, candidate_points) |
+          Cont (RollBackward rollback_point _) \<Rightarrow>
+            consumer_program ((rollback_to \<C> rollback_point), phase, candidate_points) |
+          Cont AwaitReply \<Rightarrow> \<comment> \<open>consumer is up to date\<close>
+            \<up> Done;
+            \<bottom>
+        )
+    )"
+
+definition best_intersection_point :: "chain \<Rightarrow> point list \<Rightarrow> point option" where
+  [simp]: "best_intersection_point \<C> points = (
+    let
+      intersection_points = List.set points \<inter> List.set (map hash \<C>)
+    in
+      if intersection_points = {} then None else Some (Max intersection_points)
```

Comment:
I think this layout should perhaps be changed. In the definition of `consumer_program`, you put `let`, `in`, and the equation between them all on one line and aligned the following expression with the `let`, which is also what I did in an older version of the request-response protocol, from what I remember. That said, the layout you’re using here makes sort of sense. Maybe we should keep it and adapt the `let` expression above accordingly. After decades of functional programming, I still haven’t come up with a satisfying layout for `let` expressions. :sweat_smile: